paramiko/pkey.py

```diff
@@ -31,11 +32,87 @@
 
 from paramiko import util
 from paramiko.common import o600
-from paramiko.py3compat import u, encodebytes, decodebytes, b, string_types
+from paramiko.py3compat import (
+    u,
+    encodebytes,
+    decodebytes,
+    b,
+    string_types,
+    StringIO,
+)
 from paramiko.ssh_exception import SSHException, PasswordRequiredException
 from paramiko.message import Message
 
 
+def _detect_private_key_type(key_str):
+    """
+    Detects the private key type and returns the PKey class that implements it.
+    Raises SSHException if unsupported.
+    """
+    match = re.match(
+        r"^-----BEGIN (OPENSSH|EC|RSA|DSA) PRIVATE KEY-----$",
+        key_str,
+        re.MULTILINE,
+    )
+    if match is None:
+        raise SSHException("private key file not supported or invalid")
+
+    keytype = match.group(1)
+
+    if keytype == "OPENSSH":
+        from paramiko import Ed25519Key
```

Comment:
this could be any type of key ... it just happens that paramiko does not implement support for the new openssh private key format for any other key type (until #1343 is merged)

Reply:
Ah thanks for pointing that out. But seems like there is nothing I can do about that right now? PR #3143 does not seem to be in a hurry. :(
Should I add a code comment here?

This PR seems to me like a step in the right direction at least, users shouldn't care about key types (unless they specifically want to). And having to specify key type in advance seems awkward for supporting the new-type keys too.

Reply:
yeah, for now a code comment is probably the best that can be done